profiling-ffi/src/string_storage.rs

```diff
@@ -0,0 +1,209 @@
+#[repr(C)]
+pub enum ManagedStringStorageNewResult {
+    Ok(ManagedStringStorage),
+    #[allow(dead_code)]
+    Err(Error),
+}
+
+#[no_mangle]
+#[must_use]
+pub unsafe extern "C" fn ddog_prof_ManagedStringStorage_new() -> ManagedStringStorageNewResult {
+    let storage = InternalManagedStringStorage::new();
+
+    ManagedStringStorageNewResult::Ok(ManagedStringStorage {
+        inner: Rc::into_raw(Rc::new(RwLock::new(storage))) as *const c_void,
+    })
+}
+
+#[no_mangle]
+/// TODO: @ivoanjo Should this take a `*mut ManagedStringStorage` like Profile APIs do?
+pub unsafe extern "C" fn ddog_prof_ManagedStringStorage_drop(storage: ManagedStringStorage) {
+    if let Ok(storage) = get_inner_string_storage(storage, false) {
+        drop(storage);
+    }
+}
+
+#[repr(C)]
+#[allow(dead_code)]
+pub enum ManagedStringStorageInternResult {
+    Ok(ManagedStringId),
+    Err(Error),
+}
+
+#[must_use]
+#[no_mangle]
+/// TODO: Consider having a variant of intern (and unintern?) that takes an array as input, instead
+/// of just a single string at a time.
+/// TODO: @ivoanjo Should this take a `*mut ManagedStringStorage` like Profile APIs do?
+pub unsafe extern "C" fn ddog_prof_ManagedStringStorage_intern(
+    storage: ManagedStringStorage,
+    string: CharSlice,
+) -> ManagedStringStorageInternResult {
+    (|| {
+        let storage = get_inner_string_storage(storage, true)?;
+
+        let string: &str = CStr::from_ptr(string.as_ptr())
+            .to_str()
+            .map_err(|_| anyhow::anyhow!("invalid utf8 string"))?;
+
+        let string_id = storage
+            .write()
+            .map_err(|_| {
+                anyhow::anyhow!("acquisition of write lock on string storage should succeed")
+            })?
+            .intern(string);
+
+        anyhow::Ok(ManagedStringId { value: string_id })
+    })()
+    .context("ddog_prof_ManagedStringStorage_intern failed")
+    .into()
+}
+
+#[no_mangle]
+/// TODO: @ivoanjo Should this take a `*mut ManagedStringStorage` like Profile APIs do?
+pub unsafe extern "C" fn ddog_prof_ManagedStringStorage_unintern(
+    storage: ManagedStringStorage,
+    id: ManagedStringId,
+) -> MaybeError {
+    if id.value == 0 {
+        return MaybeError::None;
+    }
+
+    let result = (|| {
+        let storage = get_inner_string_storage(storage, true)?;
+        storage
+            .read()
+            .map_err(|_| {
+                anyhow::anyhow!("acquisition of read lock on string storage should succeed")
+            })?
+            .unintern(id.value);
+        anyhow::Ok(())
+    })()
+    .context("ddog_prof_ManagedStringStorage_unintern failed");
+
+    match result {
+        Ok(_) => MaybeError::None,
+        Err(e) => MaybeError::Some(e.into()),
+    }
+}
+
+#[repr(C)]
+#[allow(dead_code)]
+pub enum StringWrapperResult {
+    Ok(StringWrapper),
+    Err(Error),
+}
+
+#[must_use]
+#[no_mangle]
+/// TODO: @ivoanjo It's not clear to me if the string pointer we return here is the exact one from
```

Comment:
At least as written, it's a copy which is not owned/managed by the string storage.

Reply:
Ah, you're right! I had not seen/used the `StringWrapper` before and missed it was a feature that was already added in ddcommon-ffi and that there's an explicit `ddog_StringWrapper_drop` that goes along with it.

Since I had not seen a drop in the profiling ffi I was a bit confused about how this memory was being managed.

I've pushed d4194a2e78e47bae68d565d54fdb641a4b46d2cd to fix/update this comment.